sql/api/src/main/scala/org/apache/spark/sql/internal/types/AbstractStringType.scala

```diff
@@ -42,22 +43,25 @@ case object StringTypeBinary extends AbstractStringType {
 case object StringTypeBinaryLcase extends AbstractStringType {
   override private[sql] def acceptsType(other: DataType): Boolean =
     other.isInstanceOf[StringType] && (other.asInstanceOf[StringType].supportsBinaryEquality ||
-      other.asInstanceOf[StringType].isUTF8LcaseCollation)
+      other.asInstanceOf[StringType].isUTF8LcaseCollation) &&
+      !other.asInstanceOf[StringType].usesTrimCollation
 }
 
 /**
  * Use StringTypeWithCaseAccentSensitivity for expressions supporting all collation types (binary
  * and ICU) but limited to using case and accent sensitivity specifiers.
  */
 case object StringTypeWithCaseAccentSensitivity extends AbstractStringType {
-  override private[sql] def acceptsType(other: DataType): Boolean = other.isInstanceOf[StringType]
+  override private[sql] def acceptsType(other: DataType): Boolean =
+    other.isInstanceOf[StringType] && !other.asInstanceOf[StringType].usesTrimCollation
 }
 
 /**
- * Use StringTypeNonCSAICollation for expressions supporting all possible collation types except
- * CS_AI collation types.
+ * Use StringTypeNonTrimCSAICollation for expressions supporting all possible collation types except
+ * CS_AI and Trim collation types.
  */
-case object StringTypeNonCSAICollation extends AbstractStringType {
+case object StringTypeNonTrimCSAICollation extends AbstractStringType {
```

Comment:
@uros-db @stefankandic Do we think this blocking is sufficient for trim? If I understood we are adding new type of a specifier. Shouldn't we have some way of blocking any kind of usage of this collation? For example, what if user sets default collation to trim collation and then tries to store a literal value with default string collation. Should this fail as well?

Reply:
https://github.com/apache/spark/commit/c54c017e93090a5fb2edf1b5ef029561b6387a3f blocking any kind of trim collation usage.
Idea of this pr is to have trim collation blocked for certain expressions (for the start all expressions) and then gradually whitelisting as it gets implemented.

Reply:
Yes, but whenever collation that uses trim specifier is detected we throw not supported exception, it's not related to input types. Blocked [here](https://github.com/search?q=repo%3Aapache%2Fspark+sql%2Fcore%2Fsrc%2Fmain%2Fscala%2Forg%2Fapache%2Fspark%2Fsql%2Fexecution%2FSparkSqlParser.scala+visitSetCollation&type=code)
[here](https://github.com/apache/spark/commit/c54c017e93090a5fb2edf1b5ef029561b6387a3f#diff-26fe6b511f4b8aedf5bdffae0d7c03aba69463c09fa4f881d77520125796e0fa:~:text=%2B-,val%20collationName%20%3D%20ctx.collationName.getText,%7D,-2560)
and [here](https://github.com/apache/spark/commit/c54c017e93090a5fb2edf1b5ef029561b6387a3f#diff-65ba7ae7f352e19e30537b8ad83b79629842f2d1951fb7d13396c9fe7f2fb3e2:~:text=%2B-,if%20(!SQLConf.get.trimCollationEnabled%20%26%26,%7D,-55
)

Reply:
Got it. Thanks for the clarification.